controller-server/src/main/java/com/yahoo/vespa/hosted/controller/maintenance/SystemUpgrader.java

```diff
@@ -32,34 +37,51 @@ protected void maintain() {
             return;
         }
         for (List<ZoneId> zones : controller().zoneRegistry().upgradePolicy().asList()) {
-            if (!completeUpgrade(zones, target.get())) {
+            // The order here is important. Config servers should always upgrade first
+            if (!deploy(zones, SystemApplication.configServer.with(target.get()))) {
+                break;
+            }
+            if (!deploy(zones, SystemApplication.zone.with(target.get()))) {
                 break;
             }
         }
     }
 
-    /** Returns true if upgrade of given zones is complete */
-    private boolean completeUpgrade(List<ZoneId> zones, Version version) {
+    /** Deploy application on given version. Returns true when all allocated nodes are on requested version */
+    private boolean deploy(List<ZoneId> zones, SystemApplication application) {
         boolean completed = true;
         for (ZoneId zone : zones) {
-            startUpgrade(zone, version);
-            completed = completed && !isUpgrading(zone);
+            if (!wantedVersion(zone, application.id()).equals(application.version())) {
+                log.info(String.format("Deploying %s version %s in %s", application.id(), application.version(),
+                                       zone));
+                controller().applications().deploy(application, zone);
+            }
+            completed = completed && currentVersion(zone, application.id()).equals(application.version());
```

Comment:
It's been requested (offline) that the controller tries to upgrade all (reachable) zones, but only waits for convergence among those which are `controllerUpgraded`. This will be easy to implement here at a later time!

Reply:
I can make this change once the other PR is merged.

Reply:
Great! No rush, though.